File: Multi-Way Trees/btree_implementation.py
# File: btree.py
# Program implementing B-Tree operations

class BTreeNode:
    def __init__(self, t, leaf=False):
        self.t = t
        self.leaf = leaf
        self.keys = []
        self.children = []

class BTree:
    def __init__(self, t):
        self.root = BTreeNode(t, True)
        self.t = t

    def traverse(self, node):
        for i in range(len(node.keys)):
            if not node.leaf:
                self.traverse(node.children[i])
            print(node.keys[i], end=" ")

        if not node.leaf:
            self.traverse(node.children[len(node.keys)])

    def search(self, node, k):
        i = 0
        while i < len(node.keys) and k > node.keys[i]:
            i += 1

        if i < len(node.keys) and node.keys[i] == k:
            return node

        if node.leaf:
            return None

        return self.search(node.children[i], k)

if __name__ == "__main__":

    btree = BTree(3)

    print("B-Tree structure created")

File: Multi-Way Trees/BTreeImplementation.java
// File: BTree.java
// Program implementing B-Tree operations

class BTreeNode {
    int t, n;
    int keys[];
    BTreeNode children[];
    boolean leaf;

    BTreeNode(int t, boolean leaf) {
        this.t = t;
        this.leaf = leaf;
        keys = new int[2*t-1];
        children = new BTreeNode[2*t];
        n = 0;
    }

    void traverse() {
        int i;
        for(i=0;i<n;i++) {
            if(!leaf)
                children[i].traverse();
            System.out.print(keys[i]+" ");
        }

        if(!leaf)
            children[i].traverse();
    }

    BTreeNode search(int k) {
        int i=0;

        while(i<n && k>keys[i])
            i++;

        if(i<n && keys[i]==k)
            return this;

        if(leaf)
            return null;

        return children[i].search(k);
    }
}

public class BTree {

    BTreeNode root;
    int t;

    BTree(int t) {
        this.root=null;
        this.t=t;
    }

    public void traverse() {
        if(root!=null)
            root.traverse();
    }

    public BTreeNode search(int k) {
        if(root==null)
            return null;
        return root.search(k);
    }

    public static void main(String args[]) {

        BTree t = new BTree(3);

        System.out.println("Simple B-Tree structure created");

    }
}
